### ISSUE TITLE
Task test results folder not resolving with .runsettings ResultsDirectory value

### ISSUE THREAD

**Issue by davecanderson** [microsoft/azure-pipelines-tasks#12513]
*Type*: Bug
**Task**: VsTestV2
**Version**: 2.165.3
**Agent (Hosted)**: vs2017-win2016

## Issue Description

The default _Test results folder_ `$(Agent.TempDirectory)/TestResults` is not being correctly configured by the resolved `.runsettings` [ResultsDirectory](https://docs.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2019#elements-of-a-runsettings-file) value. The path is being rewritten in the updated run settings but is not fully resolved and `TestResults` is appended to the path.

Expected:  `D:\a\_temp\_TestResults`
Actual: `D:\a\_temp\.\_TestResults\TestResults`

Removing `<ResultsDirectory />` from `.runsettings` uses the default folder  `D:\a\_temp\TestResults` as expected, see also https://stackoverflow.com/questions/60612131.

### Task Log
[build-323-log-9.txt](https://github.com/microsoft/azure-pipelines-tasks/files/4315988/build-323-log-9.txt)
```
2020-03-11T02:56:12.9643215Z ##[debug]Current phase: MS.VS.TestService.VstestConsoleAdapter.SetupPhase
2020-03-11T02:56:12.9659228Z ##[debug]SetupPhase.Run: Running necessary setup steps.
2020-03-11T02:56:12.9690076Z Provided settings file:
2020-03-11T02:56:12.9703739Z <?xml version="1.0" encoding="utf-8"?>
2020-03-11T02:56:12.9703960Z <RunSettings>
2020-03-11T02:56:12.9704133Z 	<RunConfiguration>
2020-03-11T02:56:12.9704304Z 		<MaxCpuCount>1</MaxCpuCount>
2020-03-11T02:56:12.9704505Z 		<TargetPlatform>x64</TargetPlatform>
2020-03-11T02:56:12.9704744Z     <ResultsDirectory>.\_TestResults</ResultsDirectory>
2020-03-11T02:56:12.9704988Z 	</RunConfiguration>
2020-03-11T02:56:12.9705185Z   <DeploymentEnabled>True</DeploymentEnabled>
2020-03-11T02:56:12.9705378Z </RunSettings>
2020-03-11T02:56:12.9705723Z ##[debug]<?xml version="1.0" encoding="utf-8"?>
2020-03-11T02:56:12.9706005Z ##[debug]<RunSettings>
2020-03-11T02:56:12.9706253Z ##[debug]	<RunConfiguration>
2020-03-11T02:56:12.9706544Z ##[debug]		<MaxCpuCount>1</MaxCpuCount>
2020-03-11T02:56:12.9707383Z ##[debug]		<TargetPlatform>x64</TargetPlatform>
2020-03-11T02:56:12.9707709Z ##[debug]    <ResultsDirectory>.\_TestResults</ResultsDirectory>
2020-03-11T02:56:12.9708171Z ##[debug]	</RunConfiguration>
2020-03-11T02:56:12.9708464Z ##[debug]  <DeploymentEnabled>True</DeploymentEnabled>
2020-03-11T02:56:12.9708733Z ##[debug]</RunSettings>
2020-03-11T02:56:12.9789858Z ##[debug]Code coverage node not present.
2020-03-11T02:56:12.9790204Z ##[debug]Updating the BatchSize in RunSettings to 1000
2020-03-11T02:56:12.9799555Z ##[debug]BatchSize node not found in run configuration, so adding BatchSize node
2020-03-11T02:56:12.9811126Z ##[debug]Updating ResultDirectory by editing given runsettings if needed.
2020-03-11T02:56:12.9845013Z ##[debug]UpdateIfRequiredAndParse called for TestResultsDirectoryRunsettingsParser.
2020-03-11T02:56:12.9845884Z ##[debug]CommonSettingsManager : TestResultsDirectory intialized to D:\a\_temp\.\_TestResults\TestResults
2020-03-11T02:56:12.9846368Z Updated Run Settings:
2020-03-11T02:56:12.9846509Z <RunSettings>
2020-03-11T02:56:12.9846645Z   <RunConfiguration>
2020-03-11T02:56:12.9846831Z     <MaxCpuCount>1</MaxCpuCount>
2020-03-11T02:56:12.9847026Z     <TargetPlatform>x64</TargetPlatform>
2020-03-11T02:56:12.9847300Z     <ResultsDirectory>D:\a\_temp\.\_TestResults\TestResults</ResultsDirectory>
2020-03-11T02:56:12.9847556Z     <BatchSize>1000</BatchSize>
2020-03-11T02:56:12.9847738Z   </RunConfiguration>
2020-03-11T02:56:12.9847922Z   <DeploymentEnabled>True</DeploymentEnabled>
2020-03-11T02:56:12.9848094Z </RunSettings>
2020-03-11T02:56:12.9866491Z ##[debug]PERF: SettingsUpdationPhase.UpdateSettingsAsRequired: took 20.6509 ms
```

### Task Settings
```
variables:
  BuildConfiguration: 'debug'
  BuildPlatform: 'any cpu'

steps:
- task: VSTest@2
  displayName: 'Test Assemblies'
  inputs:
    testAssemblyVer2: |
     **\$(BuildConfiguration)\**\*Tests.dll
     !**\obj\**
    runSettingsFile: '$/VTR-Processing/Tests/.runsettings'
    platform: '$(BuildPlatform)'
    configuration: '$(BuildConfiguration)'
```
### Run Settings
```
<RunSettings>
    <RunConfiguration>
        <MaxCpuCount>1</MaxCpuCount>
        <ResultsDirectory>.\_TestResults</ResultsDirectory>
        <TargetPlatform>x64</TargetPlatform>
    </RunConfiguration>
  <DeploymentEnabled>True</DeploymentEnabled>
</RunSettings>
```

**Comment by ShreyasRmsft** [comment]
You can consider using the new input we introduced to simply specifying a user defined results directory. Any agent/build variables can be used in this input field. 

Note: It will however still create a TestResults folder inside the directory you specified.

![image](https://user-images.githubusercontent.com/25194518/76506958-ff910e80-6471-11ea-97e9-746dc2c7dee4.png)

**Comment by ShreyasRmsft** [comment]
And also avoid using paths relative to your runsettings file. Provide a path that is relative one of the [well defined agent/build variables ](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml)

The reason for this is that when we override certain options in the runsettings we create the temporary one in the agent temp folder since we do not want to modify the user's source directory in any way.

**Comment by davecanderson** [comment]
For now it's easiest to simply use the default values. The issue only arose because of values in `.runsettings` from local development which have now been removed.

It's not really clear how the two settings paths should combine.  Does it make sense if the task setting just take precedent?

**Comment by ShreyasRmsft** [comment]
Yes ideally the task setting will take precedent.

The only exception to this rule is when the task setting is left at the default value "$(Agent.TempDirectory)\TestResults" in which case if there is something present in the runsettings that get honored.

The reason for this is that a while ago we did not have this input and customers were using the runsettings to change the results directory. Now with the introduction of this input we did not want to break the existing mechanism which folks were using. We also wanted to keep the default input to make it more obvious and visible where we were storing the test results by default.
